engine/src/main/java/org/terasology/rendering/nui/internal/NUIManagerInternal.java

```diff
@@ -142,6 +153,24 @@ public NUIManagerInternal(TerasologyCanvasRenderer renderer, Context context) {
             TabbingManager.tabBackInputModifier = bindsManager.getBindsConfig().getBinds(new SimpleUri("engine:tabbingModifier")).get(0);
             TabbingManager.activateInput = bindsManager.getBindsConfig().getBinds(new SimpleUri("engine:activate")).get(0);
         }
+
+        moduleEnvironment = context.get(ModuleManager.class).getEnvironment();
+
+        typeWidgetFactoryRegistry = new TypeWidgetFactoryRegistryImpl(context);
+        context.put(TypeWidgetFactoryRegistry.class, typeWidgetFactoryRegistry);
+        registerTypeWidgetFactories();
+    }
+
+    private void registerTypeWidgetFactories() {
+        for (Class<? extends TypeWidgetFactory> clazz : moduleEnvironment.getSubtypesOf(TypeWidgetFactory.class)) {
```

Comment:
I forget but I think there is a major performance problem with the subtype scanner stuff from reflection. kind of unavoidable I would guess. probably should be more careful with this class scanning type of stuff since it starts to really add to loading. has something to do with the sandbox but i'm not sure what the problem is and how it can be addressed in the short term.

Reply:
It is scans Reflections cache. 
Reflections uses structure like Map<Class,List<Class>> for Subtypes it fast enough.